Label/header customisation and better defaults

Adrian-Turjak opened the pull request:
Deals with #15 

Needs to be compiled and tested, but this should achieve what we want, or be close to it.

Adrian-Turjak commented:
@kaedroho I can't seem to get this to compile. I'm not much of a frontend developer and I haven't touched typescript before, so I may have made some mistakes.

Error is:
```
condensedinlinepanel/static/condensedinlinepanel/src/condensedinlinepanel.tsx(67,37): error TS2322: Type '{ forms: Form[]; panelLabel: string; renderCardHeader: renderCardHeaderFn; canEdit: true; canDele...' is not assignable to type 'IntrinsicAttributes & IntrinsicClassAttributes<Component<CardSetProps, ComponentState>> & { child...'.
  Type '{ forms: Form[]; panelLabel: string; renderCardHeader: renderCardHeaderFn; canEdit: true; canDele...' is not assignable to type 'CardSetProps'.
    Types of property 'formTemplate' are incompatible.
      Type 'string | undefined' is not assignable to type 'string'.
        Type 'undefined' is not assignable to type 'string'.
node_modules/@types/react-dnd/index.d.ts(83,70): error TS2314: Generic type 'Component<P, S>' requires 2 type argument(s).
node_modules/@types/react-dnd/index.d.ts(84,69): error TS2314: Generic type 'Component<P, S>' requires 2 type argument(s).
node_modules/@types/react-dnd/index.d.ts(129,66): error TS2314: Generic type 'Component<P, S>' requires 2 type argument(s).
node_modules/@types/react-dnd/index.d.ts(130,67): error TS2314: Generic type 'Component<P, S>' requires 2 type argument(s).
node_modules/@types/react/index.d.ts(165,11): error TS2559: Type 'Component<P, S>' has no properties in common with type 'ComponentLifecycle<P, S>'.
```
Which doesn't look related to my change though.

Can you have a look? All I'm trying to do is pass through the label value to the right place.

kaedroho commented:
This compiled for me, what version of Typescript do you have installed?

kaedroho commented on condensedinlinepanel/static/condensedinlinepanel/src/components/CardSet.tsx:
```diff
@@ -178,7 +179,7 @@ export class CardSet extends React.Component<CardSetProps, {}> {
                 return false;
             };
-            addButton = <button className="condensed-inline-panel__top-add-button button bicolor icon icon-plus" type="button" onClick={onClickAddButton}>Add</button>;
+            addButton = <button className="condensed-inline-panel__top-add-button button bicolor icon icon-plus" type="button" onClick={onClickAddButton}>Add {this.props.panelLabel}</button>;
```
I'm thinking it might be nicer to make the ``panelLabel`` argument replace the whole label, including "Add". This would be useful for people working in languages that might put the word for "Add" after the label.

Adrian-Turjak replied:
Works for me. I was mainly mimicking what inline panel did, but this also solves my problem.

We should then put Add as the default where the string constant is set.

Adrian-Turjak commented:
and looks like my version of typescript is `2.4.2`

I downgraded to your version, and that fared somewhat better, but for some reason the webpack bundle had a ton of random little changes.

Adrian-Turjak commented:
Ok, have played with this a bit. Gotten it to work. Missed some bits, and have decided to also tweak things a little.

Silly thing that Inline panel does is use the label for both the ADD button and the Heading. Which is weird, but whatever. I've added a heading value to this panel, but have it default to label is not set, so that it doesn't break existing usage. I'll propose a patch to fix that in core wagtail for the inlinepanel because it makes sense there too provided it still works the same until the use the heading value itself.

Will submit a new patch. I've gotten it to compile myself, but my version of webpack is different, so it adds a lot of random stuff which is unrelated. Better for you to just recompile it and test that it makes sense.

Adrian-Turjak commented:
Ok, cool. That should do it! Please compile and test yourself, and see if that works for you. :)

kaedroho commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -176,7 +177,8 @@ def bind_to_model(self, model):
             'relation_name': self.relation_name,
             'related': related,
             'panels': self.panels,
-            'heading': self.label,
```
I think we should leave the heading field called ``label`` so it's the same as Wagtail's builtin ``InlinePanel``.

kaedroho commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -176,7 +177,8 @@ def bind_to_model(self, model):
             'panels': self.panels,
-            'heading': self.label,
+            'heading': self.heading,
+            'label': self.label,
```
I think it would be really nice to make the add label default to "add <model name>" to make it consistent with the builtin InlinePanel

Adrian-Turjak replied:
Will add that in tomorrow (hopefully).

kaedroho commented:
Thanks @Adrian-Turjak!

> I've gotten it to compile myself, but my version of webpack is different, so it adds a lot of random stuff which is unrelated. Better for you to just recompile it and test that it makes sense.

Yeah, this is rather annoying.. I've noticed it happens when two people are using different OSes (I'm using Linux). I'm not really sure what we can do to fix it.

Adrian-Turjak commented:
So i'm looking at having the default be "Add {model name}" but...

I can't seem to see how if at all `InlinePanel` actually does this. And from testing it, I don't think it does. If I don't set a label it actually doesn't even seem to have the header bar:
```
InlinePanel(
     'featured_casestudies',
     label='Featured Casestudies'
),
```
    
![inline_header](https://user-images.githubusercontent.com/6404853/29904164-a0dc3e44-8e5b-11e7-87d5-5cb7ec52c449.png)

```
InlinePanel(
    'featured_casestudies',
    # label='Featured Casestudies'
),
```
![inline_noheader1](https://user-images.githubusercontent.com/6404853/29904183-b1d9bffa-8e5b-11e7-9b22-72f5c3f041ec.png)
![inline_noheader2](https://user-images.githubusercontent.com/6404853/29904184-b1f3907e-8e5b-11e7-8876-66f59c409449.png)


I'll add this to the condensed version, and potentially also to the core one since that does appear odd.

@kaedroho Can you a double check I'm not going mad and that `InlinePanel` does indeed not have a sensible default if no label is set?

Adrian-Turjak commented:
latest commit handles default cases a little better, and also deals with #5 while I'm at it.

I think we're pretty much done.

Adrian-Turjak commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -171,12 +174,17 @@ def bind_to_model(self, model):
         else:
             related = getattr(model, self.relation_name).related
 
+        related_name = camelcase_to_underscore(
```
Yes! This is much better. I just blanked on this and didn't think it through. Will follow up with that change.

Adrian-Turjak commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -154,9 +156,10 @@ def required_formsets(cls):
 
 class CondensedInlinePanel(object):
-    def __init__(self, relation_name, panels=None, label='', help_text='', min_num=None, max_num=None, card_header_from_field=None, card_header_from_js=None, card_header_from_js_safe=None):
+    def __init__(self, relation_name, panels=None, heading='', label='', help_text='', min_num=None, max_num=None, card_header_from_field=None, card_header_from_js=None, card_header_from_js_safe=None):
```
I don't see why not. It's easy and can just be one extra arg most people can ignore if not needed.

Adrian-Turjak commented:
@ababic Back to you. :)

And when/if you use this for wagtailmenus, I will be more than happy to review and test!

ababic commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -156,7 +156,7 @@ def required_formsets(cls):
+    def __init__(self, relation_name, panels=None, heading='', label='', help_text='', min_num=None, max_num=None, card_header_from_field=None, card_header_from_js=None, card_header_from_js_safe=None, new_card_header_text=""):
         self.relation_name = relation_name
         self.panels = panels
         self.heading = heading or label
```
I'm not sure using `label` for `heading` here under these conditions is very intuitive now that we define a more sensible default for heading lower down. I think we're just doing this for the sake of backwards compatibility, but I think it would be better instead to decouple the two completely and include some guidance in the release notes instead. Any thoughts?

Adrian-Turjak replied:
While I think this panel type isn't being used widely, I tend to lean towards backwards compatibility for any changes unless entirely unavoidable.

Plus, a lot of what I've ended up doing here, even with the defaults we've now added, I actually want to move to the standard wagtail inline panel, so keeping the changes here, and there in sync so the two are pretty much drop in replacements for each other (apart from card_header args) is useful. I've already started with: https://github.com/wagtail/wagtail/pull/3765

ababic replied:
I'm happy with that. Might be a good idea to mark it with a TODO comment or something though, to make that clear. Maybe somethings like:

```
# TODO: label is used below for backwards compatibility. We may want to rethink this later.
self.heading = heading or label
```

Adrian-Turjak replied:
done :)

Adrian-Turjak commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -167,24 +167,26 @@ def __init__(self, relation_name, panels=None, heading='', label='', help_text='
-            related.related_model.__name__).replace("_", " ").title()
+        heading = self.heading or _(related.related_model._meta.verbose_name_plural.capitalize())
+        new_card_header = self.new_card_header_text or _("New %s" % related.related_model._meta.verbose_name)
+        label = self.label or _('Add %s' % related.related_model._meta.verbose_name.capitalize())
```
Will drop the capitalising.

Also will redo the translations. I actually haven't played with translations much and didn't actually realise that in your examples you were calling `.format` after the `_` call. I've just had a quick read through the django docs and realised that with their example:
```
output = _('Today is %(month)s %(day)s.') % {'month': m, 'day': d}
```

I'll update with a new commit.

Although because I find it less verbose and easier to read, I tend to stick with old style string formatting unless I explicitly need new style features, so apologies in advance for that. They can take it away from me when they deprecate it (which probably won't happen). :P

ababic replied:
Yes, I must admit I still find myself using both in different places. I have a mild preference for format() recently though. Happy with this 😀

Adrian-Turjak commented:
@kaedroho I think we're good! @ababic seems happy, and it seems to work. Test yourself to make sure I haven't broken anything, but it's with you now. :)

I'll also start doing some follow pull requests to wagtail core for the inline panel itself to port some of these defaults and such over.

kaedroho commented on condensedinlinepanel/templates/condensedinlinepanel/condensedinlinepanel.js:
```diff
@@ -17,12 +21,17 @@
 
             if (value==null){
                 var value = "";
+                var needsEscaping = false;
```
No need for ``var`` here

kaedroho commented on condensedinlinepanel/templates/condensedinlinepanel/condensedinlinepanel.js:
```diff
@@ -2,10 +2,14 @@
                 {# Get the card header value from a field and escape it #}
                 var value = form.fields['{{ self.card_header_from_field|escapejs }}'];
+                if (value===null){
+                  var value = "{{ self.new_card_header }}";
```
No need for ``var`` here too

kaedroho commented on condensedinlinepanel/templates/condensedinlinepanel/condensedinlinepanel.js:
```diff
@@ -2,10 +2,14 @@
             {% if self.card_header_from_field %}
                 {# Get the card header value from a field and escape it #}
                 var value = form.fields['{{ self.card_header_from_field|escapejs }}'];
+                if (value===null){
```
What if the form is not new but the value of the specified field is ``null``? Perhaps we should check ``form.isNew`` here instead?

kaedroho commented on condensedinlinepanel/edit_handlers.py:
```diff
@@ -108,6 +109,7 @@ def get_form_extra_data(form):
             'forms': [
                 {
                     'id': i,
+                    'as_str': str(form.instance),
```
Variables on the form should be camel case. Also, I think a name should make it clear this is for the existing database instance. How about ``instanceAsStr``?

We should also add this to the ``Form`` interface in ``types.ts``

kaedroho commented:
Thanks both of you for your continued effort on this, it's looking good! Just a couple of small issues  remaining.

Adrian-Turjak commented:
@kaedroho I've added six support since I ran into a unicode error with str in py27, and hopefully also addressed all your comments.

Needs a js recompile which i've left for you, since my webpack is different. :P

Adrian-Turjak commented:
@kaedroho Any chance to recompile and review my last few changes? Want to get this merged, since then you can release a new version on pypi and I can start using it in my projects. ;)

kaedroho commented:
Apologies for the delay (been on holiday). These changes look good to me, thanks!

Adrian-Turjak commented:
@kaedroho more than welcome! And when you publish a new version to pypi I'll start using this in my current project. ;)
